### PR TITLE
Forbid some sys-calls in handle_signal

### LINKED ISSUE #1767: Restrict some abilities in handling signals
### Problem to Solve

Never send replies while processing signals: not common and not system error reply
Deny gr_source there
Deny system reservation there

### PR DESCRIPTION
**Pull request by ark0f**
Resolves #1767

### CODE REVIEW

**Comment by breathx** [comment]
Please, check that while out of allowance, system reservation rolls back and non signal produced

**Review by breathx — CHANGES_REQUESTED** [review]

File: examples/binaries/signal-entry/src/lib.rs [hunk 1]
```diff
@@ -144,6 +146,11 @@ mod wasm {
                     ))
                 );
             }
+            HandleAction::ForbiddenCallInSignal => {
+                HANDLE_SIGNAL_STATE = HandleSignalState::ForbiddenCall;
+                exec::system_reserve_gas(1_000_000_000).unwrap();
+                exec::wait();
```
[the remark this answers is not in the document]
Comment (ark0f) [reply examples/binaries/signal-entry/src/lib.rs:152]: No. State will be reverted

File: core/src/message/mod.rs [hunk 2]
```diff
@@ -137,6 +138,23 @@ impl DispatchKind {
     pub fn is_signal(&self) -> bool {
         matches!(self, Self::Signal)
     }
+
+    /// Sys-calls that are not allowed to be called for the dispatch kind.
+    pub fn forbidden_funcs(&self) -> BTreeSet<&'static str> {
+        match self {
+            DispatchKind::Signal => [
+                "gr_source",
```
Comment (breathx) [inline core/src/message/mod.rs:146]: @SabaunT someone replace this set with SysCallName maybe? Sync on it, please

File: pallets/gear/src/tests.rs [hunk 3]
```diff
@@ -6980,6 +6980,122 @@ fn system_reservation_zero_amount_panics() {
...
+        System::set_block_number(expiration - 1);
+        Gear::set_block_number((expiration - 1).try_into().unwrap());
+
+        run_to_next_block(None);
+
+        assert!(GasHandlerOf::<Test>::get_system_reserve(mid).is_err());
+
+        // check signal dispatch panicked
+        assert_eq!(MailboxOf::<Test>::iter_key(USER_1).last(), None);
+        let signal_msg_id = MessageId::generate_signal(mid, 1);
```
Comment (breathx) [inline pallets/gear/src/tests.rs:7041]: do we really need second arg for signal id generation? seems not

File: pallets/gear/src/tests.rs [hunk 4]
```diff
@@ -6980,6 +6980,122 @@ fn system_reservation_zero_amount_panics() {
...
+
+        System::set_block_number(expiration - 1);
+        Gear::set_block_number((expiration - 1).try_into().unwrap());
+
+        run_to_next_block(None);
+
+        assert!(GasHandlerOf::<Test>::get_system_reserve(mid).is_err());
+
+        // check signal dispatch panicked
+        assert_eq!(MailboxOf::<Test>::iter_key(USER_1).last(), None);
```
Comment (breathx) [inline pallets/gear/src/tests.rs:7040]:
```suggestion
        assert!(MailboxOf::<Test>::is_empty(&USER_1));
```

File: pallets/gear/src/tests.rs [hunk 5]
```diff
@@ -6980,6 +6980,122 @@ fn system_reservation_zero_amount_panics() {
...
+        });
+
+        let expiration = expiration.unwrap();
+
+        System::set_block_number(expiration - 1);
+        Gear::set_block_number((expiration - 1).try_into().unwrap());
+
+        run_to_next_block(None);
+
+        assert!(GasHandlerOf::<Test>::get_system_reserve(mid).is_err());
```
Comment (breathx) [inline pallets/gear/src/tests.rs:7037]: yeah sorry again, too many logic on signals and reservations:
do we have tests showing free of funds once system reservation disappear?

File: pallets/gear/src/tests.rs [hunk 6]
```diff
@@ -6980,6 +6980,122 @@ fn system_reservation_zero_amount_panics() {
...
+            HandleAction::Simple.encode(),
+            min_limit,
+            0,
+        ));
+
+        let mid = get_last_message_id();
+
+        run_to_block(3, Some(min_limit - 1));
+
+        assert_eq!(GasHandlerOf::<Test>::get_system_reserve(mid), Ok(0));
```
Comment (breathx) [inline pallets/gear/src/tests.rs:7095]: why is it zeroed?

Comment (breathx) [inline pallets/gear/src/tests.rs:7095]: for me seems also reasonable to remove reservation once it's zeroed

**Review by breathx — CHANGES_REQUESTED** [review]

**Comment by breathx** [comment]
It seems we also need test that if signal triggered it won't send error reply on `core_processor::precharge/prepare` fns Notes generation? Recheck manually also

File: pallets/gear/src/tests.rs (hunk 6 above)
Comment (ark0f) [reply pallets/gear/src/tests.rs:7095]: Because message exists and it's field is just zero

Comment (ark0f) [reply pallets/gear/src/tests.rs:7095]: There no reservation. That's why it's zero

File: core/src/message/mod.rs (hunk 2 above)
Comment (ark0f) [reply core/src/message/mod.rs:146]: Will be done in another PR as it requires some refactoring

File: pallets/gear/src/tests.rs [hunk 7]
```diff
@@ -6564,6 +6564,135 @@ fn signal_recursion_not_occurs() {
...
+            System::events()
+                .into_iter()
+                .filter(|e| {
+                    matches!(
+                        e.event,
+                        MockRuntimeEvent::Gear(Event::UserMessageSent { .. })
+                    )
+                })
+                .count(),
+            2 // reply from program + reply to user because of panic
```
Comment (breathx) [inline pallets/gear/src/tests.rs:6624]: two replies from the same context? I'd expand docs

**Review by breathx — APPROVED** [review]

**Pull request merged**